CanonicalABI.md

```diff
@@ -0,0 +1,1082 @@
+    # and as an integer handle when receiving as part of an export.
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        [i32, i32, i32]
+      else:
+        [i32]
+
+# Determines the core wasm types used to represent a variant's cases. This
+# overlays various cases into the same list of types, so the length of types is
+# the length of the longest case, and types are "unified" to a common
+# representation to ensure there's enough bits to hold enough values for all
+# cases.
+def flatten_cases(direction, cases):
+  result = []
+  for case in cases:
+    if case:
+      tys = flatten(direction, case)
+      for i, ty in enumerate(tys):
+        if i < len(result):
+          result[i] = unify(result[i], ty)
+        else:
+          result.push(ty)
+  return result
+
+def unify(a, b):
+  match (a, b)
+    (i64, _) |
+    (_, i64) |
+    (i32, f64) |
+    (f64, i32) => i64,
+
+    (i32, i32) |
+    (i32, f32) |
+    (f31, i32) => i32,
+
+    (f32, f32) => f32,
+    (f64, f64) | (f32, f64) | (f64, i32) => f64,
+```
+
+The only difference in exports/imports currently is that `push-buffer` and
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics"
+will be used to operate on the handle.
+
+## Lowering
+
+The `lower` function here takes three parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lowering is happening.
+* `dst` - this is an instance where the value is being lowered into. This could
+  be a WebAssembly instance or the host itself.
+* `value` - the interface types value that's being lowered
+
+This defines how an interface types value is encoded into linear memory and
+WebAssembly values.
+
+```
+def lower(direction, dst, value):
+  match value
+    f32(val) => [val as wasm_f32]
+    f64(val) => [val as wasm_f64]
+
+    # sign-extended
+    s8(val) => [val as wasm_i32]
+    s16(val) => [val as wasm_i32]
+
+    # zero-extended
+    u8(val) => [val as wasm_i32]
+    u16(val) => [val as wasm_i32]
+
+    s32(val) => [val as wasm_i32]
+    u32(val) => [val as wasm_i32]
+    s64(val) => [val as wasm_i64]
+    u64(val) => [val as wasm_i64]
+
+    char(val) => [val as wasm_i32]
+
+    string(val) => {
+      base = dst.canonical_abi_malloc(val.len(), 1)
+      encode_utf8_string(dst, base, val)
+      [base as wasm_i32, val.utf8_len() as wasm_i32]
+    }
+
+    # List are represented by their pointer/length as 32-bit integers. Note that
+    # the `canonical_abi_malloc` can be elided if `dst` is the host itself since
+    # the host can retain the raw pointers into the wasm module's memory if it
+    # likes.
+    list<$ty>(val) => {
+      base = dst.canonical_abi_malloc(val.len() * size(direction, $ty), align(direction, $ty))
```

Comment:
ditto for `realloc`

Reply:
I included `realloc` in the section about intrinsics, but do you think it also needs to be mentioned around the lifting/lowering?